### ISSUE TITLE
prefer-const rule seems to be confused by expressions in a while condition

### ISSUE THREAD

**Issue by julienw** [eslint/eslint#5024]
**What version are you using?**
2.0.0-beta.1

**What did you do?**
I use this code:
```javascript
  let parsed;
  let result = '';
  while ((parsed = re.exec(str)) !== null) {
    const [, count, letter] = parsed;
    result += letter.repeat(count);
  }
```
With the "prefer-const" rule, I get `'parsed' is never modified, use 'const' instead. (prefer-const)` which seems incorrect, because it is indeed modified in the while condition.

**Comment by eslintbot** [comment]
@julienw Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by julienw** [comment]
Exactly:
```
/home/julien/perso/git/woggle/src/js/develop.js
  3:7  error  'parsed' is never modified, use 'const' instead  prefer-const
```

**Comment by mysticatea** [comment]
Thank you for this issue.

Ah right, it's a bug.
I will fix this today (JST).

Currently, `prefer-const` is warning variables below:

- It's declared by `let`.
- It has only 1 `WriteReference`.
- There are no `ReadReference` outside of the scope that the `WriteReference` exists.
- It's not at `ForStatement.init`.

**Comment by egoroof** [comment]
I have the same issue with beta.2 with following code:
```js
let id;

function foo() {
    if (typeof id === 'undefined') {
        return;
    }
    id = setInterval(() => {
        // some stuff
    }, 250);
}

foo();
```
Does the fix affect to my code too?

**Comment by nzakas** [comment]
Please try the master branch and check.

**Comment by egoroof** [comment]
I've checked https://github.com/eslint/eslint/commit/6433b7b4372468d3470d32f6c2285fa9f06e24fa and problem with my code isn't resolved. The output is `1:5  error  'id' is never modified, use 'const' instead  prefer-const`. Should I create a new issue for this?

**Comment by julienw** [comment]
yes please :)
